src/imitation/algorithms/preference_comparisons.py

```diff
@@ -849,7 +888,7 @@ def train(self, total_timesteps: int, total_comparisons: int) -> Mapping[str, An
             # the desired total number of comparisons, we collect the remainder
             # right at the beginning to pretrain the reward model slightly
             if i == 0:
-                num_pairs += extra_comparisons
+                num_pairs += extra_comparisons + initial_comparisons
```

Comment:
Is this enough? Gathers a bunch of data, but only does a single reward model update AFAICT. Might need an `initial_updates` or make that proportional to the standard number of `self.comparisons_per_iteration`.

Reply:
Yeah, agreed. I've just made the number of reward training epochs proportional to the number of new comparisons now, wanted to avoid yet another hyperparameter. If we decide we need something like `initial_updates` after all, then maybe we should just implement a schedule that regulates the amount of reward vs policy training (IIRC, the DRLHP paper decays the amount of reward training over time).

My implementation with `epoch_multiplier` seems a bit complicated, maybe we should just pass `epochs` to `reward_trainer.train()`, rather than making it a constructor argument. OTOH, the entire point of having abstract base classes was that the main preference comparison trainer shouldn't make many assumptions about the individual components, and an epoch multiplier is a weaker assumption than an absolute number of epochs.

We could consider just getting rid of the ABC for the reward trainer, or at least making a few more assumptions (such as having epochs over a fixed training set). I think that would simplify the code, but if we'll predictably want to have very different types of reward trainers at some point, I would leave things as they are.

Reply:
I think I misunderstood what the code did previously: I'd assumed we just treated the PreferenceDataset as one big batch and did a single update, but I see now that we do batch it (which is reasonable). I'd have actually thought a priori that a fixed number of epochs for the initial comparison should be fine. Maybe it should be a different number to for non-initial comparisons (on the basis that it's a different size dataset), though.

It's an interesting question how the # of epochs should scale with dataset size. Having it be proportional to size of dataset seems not quite right: it'd mean the number of training updates is quadratic in the number of initial comparisons, as we train for more batches and more epochs the more data we have. I actually usually see the opposite -- when training on very big datasets like with unsupervised natural language models, people will often train for 1 epoch, or less (i.e. only train on random subset of data).

Was adding `epoch_multilpier` necessary to get good performance? If not, inclined to scrap it (sorry for the miscommunication). If so, then I think we should set it to some fixed value for initial vs remaining training (a new hyperparameter, which is annoying, but is OK if it has a sensible default), or implement a schedule as you suggest.

Reply:
Since it's been a while, I'm unfortunately not sure whether `epoch_multiplier` was necessary. For now, I've reverted to the original version (i.e. same number of epochs no matter the dataset size). However, I think it makes sense to implement a schedule since this is what the DRLHP paper uses (and we should probably see if replicating that helps with our issues). So I've left the `epoch_multiplier` argument in place in the reward trainer, I'm just no longer using it for now. If you'd rather I remove it for now and re-introduce it when implementing a schedule, I can do that too

Reply:
How are you planning to use `epoch_multiplier` in the future? I see the DRLHP paper uses a schedule for the rate of labeling described in appendix A:

> In the Atari domain we also pretrain the reward predictor for 200 epochs before beginning RL training, to reduce the likelihood of irreversibly learning a bad policy based on an untrained predictor. For the rest of training, labels are fed in at a rate decaying inversely with the number of timesteps; after twice as many timesteps have elapsed, we answer about
half as many queries per unit time.

But this seems to be about the labels added to the dataset, not the number of training epochs, although I could imagine them having a similar effect.

They also use a replay buffer which I think we don't use?

> We maintain a buffer of only the last 3,000 labels and loop over this buffer continuously; this is to ensure that the predictor gives enough weight to new labels (which can represent a shift in distribution) when the total number of labels becomes large

Reply:
You're right, I misremembered what they use a schedule for. I don't have a good intuition for which aspects will be important to make preference comparisons works, so it might be best to stay very close to the DRLHP paper for now. That means I've added a new parameter that specifies the `epoch_multiplier` for pretaining (i.e. we can use that to train for 200 epochs in the first round but then train less during all remaining iterations). Can easily generalize this to a schedule if we ever need to.

I'll limit the size of our dataset and introduce the preference collection schedule in another PR.